mitodl/mitxonline
# 1718 improve hubspot bulk update database queries

# What are the relevant tickets?
https://github.com/mitodl/mitxonline/issues/1718

# Description (What does it do?)
This PR makes a few improvements to the current HubSpot api in MITx Online, specifically:
1. Retrieves all objects (User, Order, Product, Line) in one database query when making batch update/create calls to HubSpot.
2. Passes objects (User, Order, Product, Line) as parameters when possible to avoid repetitive database queries.

# How can this be tested?
The steps below just ensure that everything still functions as expected.  To really test whether this PR will improve the speed of bulk create/update requests to HubSpot, we will need to test this in RC.
1. Configure MITx Online with HubSpot (ask if you need assistance with this).
2. Create a new product through Django Admin.
3. Register a new user in MITx Online and verify that a new contact is created in HubSpot.
4. Update the user through MITx Online (change their name or other attribute) and verify that the contact is updated in HubSpot.
5. Enroll the user into a course for free and verify that a new Deal is created in HubSpot.  Verify that the Deal includes a Line item matching the user's enrollment.
6. Upgrade the user's enrollment and verify that Deal in HubSpot is closed and the Line item is updated to reflect the enrollment mode.
7. Unenroll the user from the course run and verify that the Line item in HubSpot is updated to reflect the enrollment mode and change status.
8. Manually update one of the user's attributes through Django Admin.
9. Run `./manage.py sync_db_to_hubspot --users update`
10. Verify that the Contact in HubSpot is updated to reflect the change from step 8.
11. Manually update the user's enrollment mode through Django Admin.
12. Run `./manage.py sync_db_to_hubspot --lines update`
13. Verify that the Line in HubSpot is updated to reflect the change from step 11.
14. Manually create a new user through Django Admin.
15. Run `./manage.py sync_db_to_hubspot --users create`
16. Verify that a new Contact is created in HubSpot for the user from step 14.
17. Manually create a new course run enrollment and Order for the user from step 14.  The Order should reference the same course as the course run enrollment.  The Order's product should reference the product created in step 2.  The rest of the information can be random.
18. Run `./manage.py sync_db_to_hubspot --deals create`
19. Verify that a new Deal is created in HubSpot with a Line item that matches the user's course run enrollment.
